crypto/ml_dsa/ml_dsa_encoders.c

```diff
@@ -0,0 +1,995 @@
+    if (ctx == NULL)
+        goto err;
+    if (!PACKET_buf_init(&pkt, in, in_len)
+            || !PACKET_copy_bytes(&pkt, key->rho, sizeof(key->rho)))
+        goto err;
+    for (i = 0; i < key->t1.num_poly; i++)
+        if (!poly_decode_10_bits(key->t1.poly + i, &pkt))
+            goto err;
+
+    /* cache the hash of the encoded public key */
+    if (!shake_xof(ctx, key->shake256_md, in, in_len, key->tr, sizeof(key->tr)))
+        goto err;
+
+    key->pub_encoding = OPENSSL_memdup(in, in_len);
+    ret = (key->pub_encoding != NULL);
+err:
+    EVP_MD_CTX_free(ctx);
+    return ret;
+}
+
+/*
+ * @brief Encode the private key as an array of bytes.
+ * See FIPS 204, Algorithm 24, skEncode().
+ *
+ * @param key A key object containing private key values. The encoded private
+ *            key data is stored in this key.
+ * @returns 1 if the private key was encoded successfully or 0 otherwise.
+ */
+int ossl_ml_dsa_sk_encode(ML_DSA_KEY *key)
+{
+    int ret = 0;
+    const ML_DSA_PARAMS *params = key->params;
+    size_t i, written = 0, k = params->k, l = params->l;
+    ENCODE_FN *encode_fn;
+    size_t enc_len = params->sk_len;
+    const POLY *t0 = key->t0.poly;
+    WPACKET pkt;
+    uint8_t *enc = OPENSSL_malloc(enc_len);
+
+    if (enc == NULL)
+        return 0;
+
+    /* eta is the range of private key coefficients (-eta...eta) */
+    if (params->eta == ML_DSA_ETA_4)
+        encode_fn = poly_encode_signed_4;
+    else
+        encode_fn = poly_encode_signed_2;
+
+    if (!WPACKET_init_static_len(&pkt, enc, enc_len, 0)
+            || !WPACKET_memcpy(&pkt, key->rho, sizeof(key->rho))
+            || !WPACKET_memcpy(&pkt, key->K, sizeof(key->K))
+            || !WPACKET_memcpy(&pkt, key->tr, sizeof(key->tr)))
+        goto err;
+    for (i = 0; i < l; ++i)
+        if (!encode_fn(key->s1.poly + i, &pkt))
+            goto err;
+    for (i = 0; i < k; ++i)
+        if (!encode_fn(key->s2.poly + i, &pkt))
+            goto err;
+    for (i = 0; i < k; ++i)
+        if (!poly_encode_signed_two_to_power_12(t0++, &pkt))
+            goto err;
+    if (!WPACKET_get_total_written(&pkt, &written)
+            || written != enc_len)
+        goto err;
+    OPENSSL_clear_free(key->priv_encoding, enc_len);
+    key->priv_encoding = enc;
+    ret = 1;
+err:
+    WPACKET_finish(&pkt);
+    if (ret == 0)
+        OPENSSL_clear_free(enc, enc_len);
+    return ret;
+}
+
+/*
+ * @brief The reverse of ossl_ml_dsa_sk_encode().
+ * See FIPS 204, Algorithm 24, skDecode().
+ *
+ * @param in An encoded private key.
+ * @param in_len The size of |in|
+ * @param key A key object to store the decoded private key into.
+ *
+ * @returns 1 if the private key was decoded successfully or 0 otherwise.
+ */
+int ossl_ml_dsa_sk_decode(ML_DSA_KEY *key, const uint8_t *in, size_t in_len)
+{
+    DECODE_FN *decode_fn;
+    const ML_DSA_PARAMS *params = key->params;
+    size_t i, k = params->k, l = params->l;
+    PACKET pkt;
+
+    if (key->priv_encoding != NULL)
```

Comment:
This however, allows adding a private key to a public key, which is still a key mutation. The test should instead be that public key is not present.

Reply:
This ties back to the fromdata() where we disagree. I do the public key load first, then the private key load if it exists. If the public key isnt loaded then the private key part generates it.

Reply:
changed it to check both.